app/src/main/java/com/amaze/filemanager/utils/glide/CloudIconModelLoader.java

```diff
@@ -0,0 +1,41 @@
+package com.amaze.filemanager.utils.glide;
+
+import android.content.Context;
+import android.support.annotation.Nullable;
+
+import com.amaze.filemanager.database.CloudHandler;
+import com.bumptech.glide.load.Options;
+import com.bumptech.glide.load.model.ModelLoader;
+import com.bumptech.glide.signature.ObjectKey;
+
+/**
+ * Created by Vishal Nehra on 3/27/2018.
+ */
+
+public class CloudIconModelLoader implements ModelLoader<String, String> {
+
+    private Context context;
+
+    public CloudIconModelLoader(Context context) {
+        this.context = context;
+    }
+
+    @Nullable
+    @Override
+    public LoadData<String> buildLoadData(String s, int width, int height, Options options) {
+        // we put key as current time since we're not disk caching the images for cloud,
+        // as there is no way to differentiate input streams returned by different cloud services
+        // for future instances and they don't expose concrete paths either
+        return new LoadData(new ObjectKey(System.currentTimeMillis()), new CloudIconDataFetcher(context, s));
```

Comment:
`new ObjectKey(System.currentTimeMillis())`?!! This is be a key for caching, and I already told you about this error.

Reply:
Left a comment about it there.

Reply:
You [said](https://github.com/TeamAmaze/AmazeFileManager/issues/893#issuecomment-388588538): 

> Thank you, your last advice about loadData handling all the heavy work for icon loading really helped as I seem to have skipped it while reading documentation.
> I've made a PR which should fix this issue.

Not addressing caching.

Reply:
Comment above this line of code in program.

Reply:
I'm blind :smile:. 

> we put key as current time since we're not disk caching the images for cloud

We are memory caching.

> as there is no way to differentiate input streams returned by different cloud services

Can't you use the string as a unique identifier? Is it random? Doesn't the server provide a hash?

Reply:
No, and no.

Reply:
Doesn't the server provide a hash?

Reply:
I am not familiar with CloudRail in any way, but I noticed that in `CloudStorage` there is a way of getting metadata:

```java
CloudMetaData getMetadata(String var1);
```

In `CloudMetaData` there is:
```java
    private String path;
    private String name;
    private int size;
    private boolean folder;
    private Long modifiedAt;
    private ImageMetaData imageMetaData;
    [...]
    public int hashCode() {
        [...]
        return result;
    }

```
That hash should be unique enough.